### PR TITLE
Add RandAlpaka

### PR DESCRIPTION
**Pull request by jkelling**
### Features Added:
* Portable implementations of distributions (incl. Box-Muller for Gaussian)
* Default to TinyMersenneTwister, which is plain C and thus callable in
  many device contexts albeit maybe not most efficient on GPU

### Relations
* Required by https://github.com/alpaka-group/alpaka/pull/1308 to provide a generic PRNG that can be compiled for accelerator
* Needs to be taken into account by https://github.com/alpaka-group/alpaka/pull/1319 in so far as to not break https://github.com/alpaka-group/alpaka/pull/1308 .
* This may help to address https://github.com/alpaka-group/alpaka/issues/1441 by adding a simple Box-Muller transform.

# ToDo

- [x] merge after #1446
- [x] clean-up branch

### CODE REVIEW

File: include/alpaka/rand/RandAlpaka.hpp
Comment (bernhardmgruber) [inline include/alpaka/rand/RandAlpaka.hpp]: I don't like the name `RandAlpaka`. Please either call it `RandTinyMersenneTwister` to indicate what implemenatation it is, or if that does not matter we could call it `RandDefault`. But I prefer the first one.

This would also affect the header names unfortunately.

Comment (jkelling) [reply include/alpaka/rand/RandAlpaka.hpp]: As #1319 is merged now: @sliwowitz do you think it would be simple to use the new Philox instead of the TinyMT?

We can rename to RandDefault.

Comment (sliwowitz) [reply include/alpaka/rand/RandAlpaka.hpp]: I think it should be possible. I'll give it a try and report back here.

**Comment by bernhardmgruber** [comment]
Please rebase!

**Comment by jkelling** [comment]
I would like to look into using the Philox instead of TinyMT before this is merged.The rebase will happen in this process.

**Comment by jkelling** [comment]
* Renamed to `RandDefault`
* Uses Philox now
* Fix two signedness issues in Philox (`-Werror,-Wsign-compare,-Wsign-conversion`)

**Comment by jkelling** [comment]
Can be merged after CI is through.

**Comment by sliwowitz** [comment]
This PR removed  the STL generators - switching to Philox also removes TinyMT from the alpaka::rand API doesn't it?

**Comment by jkelling** [comment]
> This PR removed the STL generators - switching to Philox also removes TinyMT from the alpaka::rand API doesn't it?

No. It only adds `RandDefault` which is a handle for the portable fallback RNG and random distributions. It does not touch `RandStdLib` or the TinyMT. The first version was using TinMT (the CPU version) as the fallback RNG, the change is to use Philox instead.

**Comment by sliwowitz** [comment]
Oh. I thought you replaced the RandStdLib file, but now I see you're adding RandDefault as a default instead.

**Comment by sliwowitz** [comment]
OK. A switch to Philox as seems like a good idea because then we have identical implementation for all `Acc`s by default. I'd like to check my implementations performance before we do a release, but that's no blocker for this PR.

**Review by sliwowitz — DISMISSED** [review]

**Comment by jkelling** [comment]
> OK. A switch to Philox as seems like a good idea because then we have identical implementation for all `Acc`s by default. I'd like to check my implementations performance before we do a release, but that's no blocker for this PR.

This is not across all `Acc`s. It is currently only used by `AccOacc` and `AccOmp5`, because these require the portability. Other have backend-specific RNG, like something based onf curan/hiprand for CUDA/HIP and `RandStdLib` for the CPU backends.

Having all backend use `RandDefault` as their default is an `alpaka::rand` redesign issue, which is outside the scope of this PR.

File: include/alpaka/rand/RandDefault.hpp
Comment (bernhardmgruber) [inline include/alpaka/rand/RandDefault.hpp]:
```suggestion
                        using type = std::uint32_t;
```
In more places as well.

File: include/alpaka/rand/RandDefault.hpp
Comment (bernhardmgruber) [inline include/alpaka/rand/RandDefault.hpp]:
```suggestion
                            ret ^= (static_cast<T>(engine()) << (sizeof(BitsT) * CHAR_BIT * a));
```

File: include/alpaka/rand/RandDefault.hpp
Comment (bernhardmgruber) [inline include/alpaka/rand/RandDefault.hpp:58]: Since further down in the code you assume that `T` is an integral type, consider adding a `static_assert` for that.

Comment (bernhardmgruber) [reply include/alpaka/rand/RandDefault.hpp:58]: Same for `UniformReal`.

File: include/alpaka/rand/RandAlpaka.hpp
Comment (psychocoderHPC) [reply include/alpaka/rand/RandAlpaka.hpp]: Please check the register-footprint of `Philox` before making it the default. As I know from @sliwowitz `Philox` is generating 4 numbers within one call so it could be generating rng number that could have a larger footprint.
A large memory footprint means mostly slower kernel execution.

File: include/alpaka/rand/RandDefault.hpp
Comment (bernhardmgruber) [inline include/alpaka/rand/RandDefault.hpp]: Can you call `std::isnan` in device code? I am pretty sure you must call an alpaka math routine here. And if it does not exist, create it.

File: include/alpaka/rand/RandAlpaka.hpp
Comment (jkelling) [reply include/alpaka/rand/RandAlpaka.hpp]: It is a fallback, not yet supposed to be the default for all `Acc`s. Using it as default for other backends should be discussed separately.

File: include/alpaka/rand/RandDefault.hpp
Comment (jkelling) [reply include/alpaka/rand/RandDefault.hpp]: Yes. This works with NVHPC and also OpenMP traget compilers aim to allow math functions form the `std` namespace. It also work in CUDA.

But your are right, that we may not want to rely on that. Adding this to `alpaka::math` should also come with it friends like `isfinite`, `isinf`, ... (six in total). I should put this in front in a separate PR, then.

File: include/alpaka/math/MathStdLib.hpp
Comment (psychocoderHPC) [inline include/alpaka/math/MathStdLib.hpp]: Please provide the first commit with new math functions within a separate PR.
This improvement is very useful but sneaked in via an independent RNG PR. It is not possible to write a changelog if this commit stays in this PR.

**Review by psychocoderHPC — CHANGES_REQUESTED** [review]

File: include/alpaka/rand/RandDefault.hpp
Comment (psychocoderHPC) [reply include/alpaka/rand/RandDefault.hpp]: Yes please move this into a separate PR :heart:

File: include/alpaka/math/MathStdLib.hpp
Comment (jkelling) [reply include/alpaka/math/MathStdLib.hpp]: https://github.com/alpaka-group/alpaka/pull/1446

File: include/alpaka/rand/RandDefault.hpp
Comment (psychocoderHPC) [inline include/alpaka/rand/RandDefault.hpp:122]: IMO the copy operator of `NormalReal` should be deleted to avoid that the user is storing this object in global data. Maybe adding a comment to oxygen of this class could also help.

File: include/alpaka/rand/RandDefault.hpp
Comment (psychocoderHPC) [inline include/alpaka/rand/RandDefault.hpp]: Why is `TwoPi` starting with upper case latter and all other variables with lower?

**Review by psychocoderHPC — CHANGES_REQUESTED** [review]

File: include/alpaka/rand/RandDefault.hpp
Comment (bernhardmgruber) [reply include/alpaka/rand/RandDefault.hpp:122]: Technically, copy and move are already implicitly deleted because `NormalReal` contains a reference.

Comment (jkelling) [reply include/alpaka/rand/RandDefault.hpp:122]: @bernhardmgruber I do not think this is the case, at least with gcc 11:
```c++
#include <iostream>

struct A
{
	int& a;

	A(int &a) : a(a) {}
};

int main()
{
	int i = 23;
	A a(i);
	A b(a);
	a.a = 42;
	std::cout << b.a << std::endl;
}
```
Compiles without warnings `g++ --std=c++17 -Wall -pedantic` and prints what one would expect.

The compiler just turns the reference into a pointer. Using a reference there instead of pointer is just syntactic sugar, if the compiler does not optimize everything out.

Comment (psychocoderHPC) [reply include/alpaka/rand/RandDefault.hpp:122]: Should we explicitly delete this copy and/or move constructor?

Comment (jkelling) [reply include/alpaka/rand/RandDefault.hpp:122]: I am not sure how this really keeps the user form storing it in global memory and sharing it (good point about the comment).

Deleting copy would avoid someone copying the cache and possibly getting two times the same value. Not other distribution has deleted copies, so this would make this one different. If the copy is not deleted, it should not be trivial and clear the cache of the copy, but this would add overhead which is not there for other distributions. As it stands deleting copy could break generic code, so I would avoid this unless all distributions' copies are deleted (same for move.)

Comment (bernhardmgruber) [reply include/alpaka/rand/RandDefault.hpp:122]: Ok, let's be precise: the copy and move assignment operators are deleted. The constructors should not be impacted.

> The compiler just turns the reference into a pointer.

This happens in the codegen phase. But syntactically, you still have a reference which restricts the surrounding class.

Comment (jkelling) [reply include/alpaka/rand/RandDefault.hpp:122]: Deleting both move and copy breaks the tests unless we move to C++17 where copy elision is mandatory for return values.

Comment (bernhardmgruber) [reply include/alpaka/rand/RandDefault.hpp:122]: > Should we explicitly delete this copy and/or move constructor?

Even if you delete them, you could still placement new a `NormalReal` into a buffer. And IMO there is nothing wrong with copying/moving that class around. So I would even change the reference member to a pointer member, so the class can be copied/moved. But we should document that this class keeps a reference to the passed accelerator and that an instance should not outlive a kernel invocation.

Comment (jkelling) [reply include/alpaka/rand/RandDefault.hpp:122]: @bernhardmgruber was referring to the copy/move assignment operators there, those are implicitly deleted.

File: include/alpaka/rand/RandDefault.hpp
Comment (jkelling) [reply include/alpaka/rand/RandDefault.hpp]: Fixed.

File: include/alpaka/rand/RandDefault.hpp
Comment (jkelling) [reply include/alpaka/rand/RandDefault.hpp:122]: Pushed https://github.com/alpaka-group/alpaka/pull/1444/commits/a8b261fe42734112e484216522c5a436fdb27191 which cleanly implements copy and move ctors.

File: include/alpaka/rand/RandDefault.hpp
Comment (bernhardmgruber) [inline include/alpaka/rand/RandDefault.hpp]: I think this behavior is surprising. If you take a copy of a `NormalReal` and then ask for the next number on the old and new instance, you will get different results.

I would prefer to have the default copy ctor here.

File: include/alpaka/rand/RandDefault.hpp
Comment (bernhardmgruber) [inline include/alpaka/rand/RandDefault.hpp]: I don't think there is actually a need to define move operations for this class. What should the semantic be of moving a distribution? And why should that influence the moved from distribution? What is the need to destroy the moved from object? We don't get any efficiency by enabling move operations.

File: include/alpaka/rand/RandDefault.hpp
Comment (bernhardmgruber) [reply include/alpaka/rand/RandDefault.hpp:122]: > Deleting both move and copy breaks the tests unless we move to C++17 where copy elision is mandatory for return values.

Please also turn the stored accelerator object into a pointer, so the class can be freely copied. That should make the tests work.

Comment (jkelling) [reply include/alpaka/rand/RandDefault.hpp:122]: > And IMO there is nothing wrong with copying/moving that class around.
> So I would even change the reference member to a pointer member, so the class can be copied/moved.

There is no harm as such, but for performance it would be best to pass a reference to it around. Both copying and creating multiple instance only wastes registers on GPU. Letting an instance go out of scope and creating a new one when needed again also saves registers and is cheaper than a copy. I do not think copy and assignment of this type, or any alpaka distribution is useful, as they do not contain parameters (like min/max/sigma/mean).


> But we should document that this class keeps a reference to the passed accelerator and that an instance should not outlive a kernel invocation.

Ok. This is formally true for any object created from an `Acc`, i.e. all distributions and also random engines, this is just the only one that uses its `Acc` for something.

File: include/alpaka/rand/RandDefault.hpp
Comment (jkelling) [reply include/alpaka/rand/RandDefault.hpp]: This is only the distribution, the engine determines which numbers are generated. The expected behavior would be:
```c++
NormalReal<Acc> a(acc), b(acc);
Engine<Acc> e(acc);

inGeneral( a(e) != b(e) ) ; // because of two engine invocations

b = a;

inGeneral( a(e) != b(e) ) ; // because of two engine invocations
```

If a value was cached in `a` before `a = b` and the cache is not cleared, you break this behavior, because neither `a` nor `b` invoke the engine to ask for a new random number. This would not happen with a distribution that does not cache numbers. Caching is just an optimization, it must not interfere with the randomness.

File: include/alpaka/rand/RandDefault.hpp
Comment (jkelling) [reply include/alpaka/rand/RandDefault.hpp]: We can delete move. Though if it s there, is must be like this, for the same reason the copy should not copy the cache.

File: include/alpaka/rand/RandDefault.hpp
Comment (jkelling) [reply include/alpaka/rand/RandDefault.hpp:122]: Github issues do not appear to work when we write simultaneously. Your comments only appear to me somewhere in the middle long after I wrote something else, so I miss them. Then I get an email about a comment which turns out to be in the middle of the conversation... .

> > The compiler just turns the reference into a pointer.
> 
> This happens in the codegen phase. But syntactically, you still have a reference which restricts the surrounding class.

Yes, this is the intention: You cannot assign to a reference. `m_acc` should not change during the lifetime of the object, anything else would be a bug, hence the reference.

We could actually add assignments that do not touch the reference. See https://github.com/alpaka-group/alpaka/pull/1444/commits/31d18e754579a9ec05dd6ad71df443d293a2abb3 . This is the semantically clean way to do the assignments, though it is admittedly pointless. Note, that these object always exist withing a kernel, where there is only one `Acc` handle, i.e. you can never assign an object with a different `m_acc`.

File: include/alpaka/rand/RandDefault.hpp
Comment (jkelling) [inline include/alpaka/rand/RandDefault.hpp]: This does nothing. Basically copy should be deleted because we cannot have two distributions with the same cached value around.

File: include/alpaka/rand/RandDefault.hpp
Comment (bernhardmgruber) [reply include/alpaka/rand/RandDefault.hpp]: Ok, that's right! I could still make the case tough that you could copy the engine as well and expect the results to be the same, but maybe that is not a relevant use case. Anyway, I think it's fine then.

File: include/alpaka/rand/RandDefault.hpp
Comment (bernhardmgruber) [reply include/alpaka/rand/RandDefault.hpp:122]: > Both copying and creating multiple instance only wastes registers on GPU. Letting an instance go out of scope and creating a new one when needed again also saves registers and is cheaper than a copy.

I strongly trust the optimizer to inline everything and remove all redundancies. Passing references around also costs registers and AFAIK CUDA cannot even form pointers/references into registers, so the optimizer must eliminate those anyway. So in the end it should not matter whether you copy distribution objects or pass them by references. And since the object is light-weight, I would prefer making it as regular as possible. But after this discussion, I don't care so much anymore. You are free to choose a behavior.

File: include/alpaka/rand/RandDefault.hpp
Comment (jkelling) [reply include/alpaka/rand/RandDefault.hpp]: You can either copy the cache or not, both have drawbacks, but the property I stated above is essential. If we give weight to your second case, the caching optimization not be possible. Copying the engine it a thing that might actually happen. One would expect a any distribution instance to give the same numbers on both engines, which also speaks against caching. However removing the cache and only using one of the two random numbers seems to sacrifice performance to support some very rare case. The way I see to minimize this problem is to disallow copy and move as far as possible (we only need the move ctor currently and can delete all when moving to C++17) and document the interaction of cache and engines. I pushed https://github.com/alpaka-group/alpaka/pull/1444/commits/004ae6cdd8f4a3321ecc0c0e22bf96cb9074204e along those lines.

**Comment by jkelling** [comment]
The class `NormalReal` now has a deleted copy ctor and the move ctor is marked to be deleted when alpaka moves to C++17 by a doxgen `\todo`. The move ctor clears the cache of the moved-from instance.

@bernhardmgruber @psychocoderHPC  if you approve, I will clean-up the branch before merge.

File: include/alpaka/rand/RandDefault.hpp
Comment (jkelling) [reply include/alpaka/rand/RandDefault.hpp:122]: The CUDA compiler will inline everything and remove reference function arguments in the process, as they are just the same variable. If you copy, in the case for this `NormalReal` is not redundant because it contains the cache variable which changes independently for original and copy in their respective `operator()` calls, thus the copy cannot be optimized out (I agree, that this will be different for all other distribution, because they do not have any data members). This behavior does not depend on how exactly the copy ctor or assignment behave.

Going with deleting as much as possible, to try keep the user from getting unexpected results at runtime from this class.

**Review by psychocoderHPC — DISMISSED** [review]

**Comment by jkelling** [comment]
@psychocoderHPC rebased and ready for merge after CI.

**Review by bernhardmgruber — APPROVED** [review]

**Pull request merged**
